docs/rfc/0000-application-architecture.md

```diff
@@ -0,0 +1,503 @@
+* Interoperability with other `radicle-link` applications is not defined,
+  respectively tied to the GUI-based process management
+* Key management is implemented as an afterthought
+* It is, more generally, unclear how a monolithic architecture can support
+  independent application development
+
+Going forward, we aim to diversify application development, and thus require a
+baseline architecture of how different `radicle-link` applications can interact
+with each other.
+
+
+## Overview
+
+Instead of a monolithic architecture, we admit the presence of distinct
+application concerns, lifecycles, and security constraints. Consequently, we lay
+out a service-oriented architecture, which builds on platform-specific
+primitives for process management, IPC, and credentials management. Service-
+(and thus: network-) boundaries shall be defined mainly according to the
+lifecycle of the application.
+
+Note that this matters only for a usage scenario where all installed
+`radicle-link` applications share the same user-scoped state. An alternative
+strategy would be to scope all state to the application, ie. any two
+applications would not share the same storage nor _device key_. Since this
+raises the question of how those different states are synchronised in order to
+provide a coherent view to the user, we haven't explored this approach further,
+but admit that some clever engineering on the storage layer could enable it in
+the future.
+
+We dismiss the approach of providing a local monolithic server which encompasses
+all required functionality for modularity and extensibility reasons.
+
+
+## Prerequisites
+
+### Platforms
+
+For this discussion, we divide platform targets into tiers:
+
+* **Tier 1**:
+
+  > Fully supported platform, receives testing, deployment, and system
+  > configurations (may include packaging)
+
+  * Linux (`x86_64-unknown-linux-gnu`)
+  * macOS (`x86_64-apple-darwin`)
+
+* **Tier 2**:
+
+  > Basic support ("it compiles")
+
+  * Windows
+  * Other processor architectures for tier 1 platforms
+
+* **Tier 3**:
+
+  > Not supported, but potentially worth considering when assessing architecture
+  > decisions
+
+  * Mobile (iOS, Android)
+  * Wasm
+
+
+### System Services
+
+We will rely on declarative service / process management to be provided by the
+platform, specifically "socket activation".
+
+On macOS, this is provided on macOS via [launchd]. Linux distributions
+considered as tier 1 shall, however, by limited to those based on  [systemd] for
+the time being.
+
+We also assume the presence of an agent capable of producing Ed25519 signatures
+utilising the [SSH agent protocol][ssh-agent].
+
+On macOS, the distribution `ssh-agent` integrates with the system keychain.
+Other choices include the `ssh-agent` provided by the `openssh` suite, the
+`gpg-agent`, or (on Linux systems) `gnome-keyring`.
+
+Lastly, we consider [D-Bus][dbus] for publish-subscribe messaging. D-Bus is
```

Comment:
I am myself not entirely convinced of relying on D-Bus — it is rather arcane, and has been criticised for low performance. Perhaps we should do a prototype first, also to see just how painful it is to get going on macOS.

You’re right that D-Bus can, in principle, provide socket (well, service) activation as well. It is then, however, *required* for the client to authenticate, while enforcing socket file permissions may make this unnecessary.

Reply:
It’s a user-space application

Reply:
> You’re right that D-Bus can, in principle, provide socket (well, service) activation as well. It is then, however, _required_ for the client to authenticate, while enforcing socket file permissions may make this unnecessary.

As far as I know most D-Bus installations allow the `EXTERNAL` [authentication mechanism](https://dbus.freedesktop.org/doc/dbus-specification.html#auth-mechanisms) that does use the socket to authenticate the user.

> I am myself not entirely convinced of relying on D-Bus — it is rather arcane, and has been criticised for low performance.

Alternatively, we could drop D-Bus altogether and also use service-to-service pub-sub over RPC. My main concern is about splitting this up and . If we rely on D-Bus we will have to deal with it’s arcana and I’d rather go all-in or drop it alltogether.

Reply:
> the EXTERNAL authentication mechanism 

Oh my, that looks like a pain in the neck.

> Alternatively, we could drop D-Bus altogether

Ya, maybe that's better. After all, we only need a small ringbuffer (possibly scoped by topic), and not necessarily any of the advanced functionality. There are two reasons why D-Bus could be attractive, but _possibly_ this only applies to platforms which actually integrate with it (ie. not the proprietary ones):

1. We may want processes not owned by the current session to react to events (eg. post-fetch hooks a la Fleek)
2. We may want a way to dispatch notifications which result in some user interaction (a passphrase prompt, f.ex.): say the user has `upstream` running, it may be preferable to focus its window and draw the dialog there.  Say `upstream` is **not** currently running (or not installed at all), then can some other program handle this? I _believe_ that this is a common pattern for D-Bus-enabled applications.

Arguably, both can also be solved separately (and platform-specific), so I tend to agree that we should just roll our own pub-sub daemon.